File: ruby/703.kth-largest-element-in-a-stream.rb
# frozen_string_literal: true

class KthLargest
  def initialize(max_size, nums)
    @max_size = max_size
    @nums = nums.sort.reverse[0..k]
  end

  def add(val)
    if (i = @nums.bsearch_index { |num| num < val })
      @nums.insert(i, val)
    else
      @nums.push(val)
    end

    @nums.pop if @nums.length > @max_size
    @nums.last
  end
end

File: rust/src/703.kth-largest-element-in-a-stream.rs
use std::{cmp::Reverse, collections::BinaryHeap};

struct KthLargest {
    max_size: i32,
    heap: BinaryHeap<Reverse<i32>>,
}

#[allow(dead_code)]
impl KthLargest {
    fn new(k: i32, nums: Vec<i32>) -> Self {
        let mut _self = KthLargest {
            max_size: k,
            heap: BinaryHeap::with_capacity(k as usize + 1),
        };

        for num in nums {
            _self.add(num);
        }

        _self
    }

    fn add(&mut self, val: i32) -> i32 {
        self.heap.push(Reverse(val));

        if self.heap.len() > self.max_size as usize {
            self.heap.pop();
        }

        self.heap.peek().map_or(0, |&Reverse(val)| val)
    }
}
